fuse_models/include/fuse_models/parameters/parameter_base.h

```diff
@@ -93,6 +94,22 @@ inline std::vector<size_t> loadSensorConfig(const ros::NodeHandle& nh, const std
   return {};
 }
 
+inline fuse_core::Loss::SharedPtr loadLossConfig(const ros::NodeHandle& nh, const std::string& name)
+{
+  if (!nh.hasParam(name))
+  {
+    return {};
+  }
+
+  std::string loss_type;
+  getParamRequired(nh, name + "/type", loss_type);
+
+  auto loss = fuse_core::createUniqueLoss(loss_type);
+  loss->initialize(nh.resolveName(name));
+
+  return loss;
```

Comment:
This is important:
* `fuse_core::createUniqueLoss` returns a `pluginlib::UniquePtr<T>` where `T` is `fuse_core::Loss`: https://github.com/ros/pluginlib/blob/melodic-devel/pluginlib/include/pluginlib/class_loader.hpp#L57
* Therefore, the `loss` variable would be of type `pluginlib::UniquePtr<T>`
* But the `return` type of `loadLossConfig` is `fuse_core::Loss::SharedPtr`, i.e. an `std::shared_ptr<T>` pointer
* `pluginlib::UniquePtr<T>` is actually an alias of `class_loader::ClassLoader::UniquePtr<T>`: `pluginlib::UniquePtr<T>`, which is defined as an `std::unique_ptr<T, class_loader::DeleterType<T>>;`
* The `return` statement here does an implicit cast from the `std::unique_ptr` to the `std::shared_ptr`
* I'm not sure if that passes the non-default destructor `class_loader::DeleterType<T>`
* In case the non-default destructor is not passed to the `std::shared_ptr`, I wonder if this is could explain that the node takes longer to shutdown and ROS escalates to `SIGTERM`. No crash happens though.

BTW the `class_loader::ClassLoader` class has a method to create an instance handle by an `std::shared_ptr`: https://github.com/ros/class_loader/blob/melodic-devel/include/class_loader/class_loader.hpp#L127, but curiously, the `pluginlib::ClassLoader` only has a method for an `std::unique_ptr`: https://github.com/ros/pluginlib/blob/melodic-devel/pluginlib/include/pluginlib/class_loader.hpp#L128, or for `boost::shared_ptr`: https://github.com/ros/pluginlib/blob/melodic-devel/pluginlib/include/pluginlib/class_loader.hpp#L109 :confused:

Reply:
According to this test snippet, the custom deleter is passed to the `std::shared_ptr`, so in principle everything should be fine: https://wandbox.org/permlink/A8XqXg6x6a2bRvPY

Reply:
Yeah. I've run into this unique ptr deleter function issue before. I'm sure there is a very logical and well-considered reason why shared_ptrs can hold a deleter function but unique_ptrs require an additional template argument, but it's really annoying.

Reply:
Considering this good for now and marking this conversation as resolved.